File: lab2/Shark.cs
﻿using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace lab2
{
    public class Shark : Fish, IComparable<Shark>, IEquatable<Shark>
    {
        //8
        public int CompareTo(Shark other)
        {
            if (other == null)
            {
                return 1;
            }
            if (Age != other.Age)
            {
                return Age.CompareTo(other.Age);
            }
            if (Speed != other.Speed)
            {
                return Speed.CompareTo(other.Speed);
            }
            if (Weight != other.Weight)
            {
                return Weight.CompareTo(other.Weight);
            }
            if (ColorBody != other.ColorBody)
            {
                ColorBody.Name.CompareTo(other.ColorBody.Name);
            }
            return 0;
        }

        public bool Equals(Shark other)
        {
            if (other == null)
            {
                return false;
            }
            if (Speed != other.Speed)
            {
                return false;
            }
            if (Age != other.Age)
            {
                return false;
            }
            if (Weight != other.Weight)
            {
                return false;
            }
            if (ColorBody != other.ColorBody)
            {
                return false;
            }
            return true;
        }
        public override bool Equals(object obj)
        {
            if (obj == null)
            {
                return false;
            }
            Shark sharkObj = obj as Shark;
            if (sharkObj == null)
            {
                return false;
            }
            else
            {
                return Equals(sharkObj);
            }
        }
        public override int GetHashCode()
        {
            return base.GetHashCode();
        }

        public override int Age
        {
            get
            {
                return base.Age;
            }

            protected set
            {
                if (value > 0 && value < 30)
                {
                    base.Age = value;
                }
                else
                {
                    base.Age = 20;
                }
            }
        }
        public override double Speed
        {
            get
            {
                return base.Speed;
            }

            protected set
            {
                if (value > 0 && value < 50)
                {
                    base.Speed = value;
                }
                else
                {
                    base.Speed = 30;
                }
            }
        }
        public override int Weight

        {
            get
            {
                return base.Weight;
            }

            protected set
            {
                if (value > 0 && value < 30)
                {
                    base.Weight = value;
                }
                else
                {
                    base.Weight = 25;
                }
            }
        }

        public Shark(double speed, int age, int weight, Color color)
        {
            this.Speed = speed;
            this.Age = age;
            this.ColorBody = color;
            this.Weight = weight;
            Random rand = new Random();
            startPosX = rand.Next(10, 200);
            startPosY = rand.Next(10, 200);
        }

        //6
        public Shark(string info)
        {
            string[] strs = info.Split(';');
            if (strs.Length == 4)
            {
                Speed = Convert.ToInt32(strs[0]);
                Age = Convert.ToInt32(strs[1]);
                Weight = Convert.ToInt32(strs[2]);
                ColorBody = Color.FromName(strs[3]);
            }
            Random rand = new Random();
            startPosX = rand.Next(10, 200);
            startPosY = rand.Next(10, 200);
        }
        //

        public override void moveAnimal(Graphics g)
        {
            startPosX -= ((float)Speed);
            drawAnimal(g);
        }

        public override void drawAnimal(Graphics g)
        {
            draw1Shark(g);
        }

        protected virtual void draw1Shark(Graphics g)
        {
            //Drawing1
            Pen pen = new Pen(ColorBody, 6);

            //Body
            g.DrawLine(pen, startPosX, startPosY, startPosX + Age * 3, startPosY - Weight);
            g.DrawLine(pen, startPosX + Age * 3, startPosY - Weight, startPosX + Age * 9, startPosY);
            g.DrawLine(pen, startPosX, startPosY, startPosX + Age * 3, startPosY + Weight);
            g.DrawLine(pen, startPosX + Age * 3, startPosY + Weight, startPosX + Age * 9, startPosY);

            //Fin
            g.DrawLine(pen, startPosX + Age * 3, startPosY - Weight, startPosX + Age * 6, startPosY - Weight * 2);
            g.DrawLine(pen, startPosX + Age * 6, startPosY - Weight * 2, startPosX + Age * 6, startPosY - Weight / 2);

            //Tail
            g.DrawLine(pen, startPosX + Age * 9, startPosY, startPosX + Age * 11, startPosY + Weight);
            g.DrawLine(pen, startPosX + Age * 9, startPosY, startPosX + Age * 11, startPosY - Weight);
            g.DrawLine(pen, startPosX + Age * 11, startPosY + Weight, startPosX + Age * 10, startPosY);
            g.DrawLine(pen, startPosX + Age * 11, startPosY - Weight, startPosX + Age * 10, startPosY);
        }

        //6
        public override string getInfo()
        {
            return Speed + ";" + Age + ";" + Weight + ";" + ColorBody.Name;
        }
        //
    }
}

File: src/Shark.java
import java.awt.Color;
import java.awt.Graphics;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;
import java.util.Random;

public class Shark extends Fish implements Serializable, Comparable<Shark> {
	@Override
	public int GetAge() {
		return super.GetAge();
	}

	protected void SetAge(int value) {
		if (value > 0 && value < 30) {
			super.SetAge(value);
		} else {
			super.SetAge(20);
		}
	}

	@Override
	public double GetSpeed() {

		return super.GetSpeed();
	}

	@Override
	protected void SetSpeed(int value) {
		if (value > 0 && value < 50) {
			super.SetSpeed(value);
		} else {
			super.SetSpeed(30);
		}
	}

	@Override
	public int GetWeight() {
		return super.GetWeight();
	}

	@Override
	protected void SetWeight(int value) {
		if (value > 0 && value < 30) {
			super.SetWeight(value);
		} else {
			super.SetWeight(25);
		}
	}

	public Shark(int speed, int age, int weight, Color color) {
		this.SetSpeed(speed);
		this.SetAge(age);
		this.SetColorBody(color);
		this.SetWeight(weight);
		Random rand = new Random();
		startPosX = rand.nextInt(600);
		startPosY = rand.nextInt(300);
	}

	protected void draw1Shark(Graphics g) {

		g.setColor(GetColorBody());

		// Body
		g.drawLine(startPosX, startPosY, startPosX + GetAge() * 3, startPosY
				- GetWeight());
		g.drawLine(startPosX + GetAge() * 3, startPosY - GetWeight(), startPosX
				+ GetAge() * 9, startPosY);
		g.drawLine(startPosX, startPosY, startPosX + GetAge() * 3, startPosY
				+ GetWeight());
		g.drawLine(startPosX + GetAge() * 3, startPosY + GetWeight(), startPosX
				+ GetAge() * 9, startPosY);

		// Fin
		g.drawLine(startPosX + GetAge() * 3, startPosY - GetWeight(), startPosX
				+ GetAge() * 6, startPosY - GetWeight() * 2);
		g.drawLine(startPosX + GetAge() * 6, startPosY - GetWeight() * 2,
				startPosX + GetAge() * 6, startPosY - GetWeight() / 2);

		// Tail
		g.drawLine(startPosX + GetAge() * 9, startPosY, startPosX + GetAge()
				* 11, startPosY + GetWeight());
		g.drawLine(startPosX + GetAge() * 9, startPosY, startPosX + GetAge()
				* 11, startPosY - GetWeight());
		g.drawLine(startPosX + GetAge() * 11, startPosY + GetWeight(),
				startPosX + GetAge() * 10, startPosY);
		g.drawLine(startPosX + GetAge() * 11, startPosY - GetWeight(),
				startPosX + GetAge() * 10, startPosY);

	}

	@Override
	public void drawAnimal(Graphics g) {
		// TODO Auto-generated method stub
		draw1Shark(g);
	}

	@Override
	public void moveAnimal(Graphics g) {
		// TODO Auto-generated method stub
		startPosX -= (GetSpeed());
		drawAnimal(g);
	}

	@Override
	public int compareTo(Shark other) {
		// TODO Auto-generated method stub
		if (other == null) {
			return 1;
		}
		if (GetAge() != other.GetAge()) {
			return GetAge()+"".compareTo(Integer.toString(other.GetAge()));
		}
		if (GetSpeed() != other.GetSpeed()) {
			return (int)GetSpeed()+"".compareTo(Double.toString(other.GetSpeed()));
		}
		if (GetWeight() != other.GetWeight() ) {
			return GetWeight()+"".compareTo(Integer.toString(other.GetWeight()) );
		}
		if (GetColorBody() != other.GetColorBody()) {
			return GetColorBody().toString().compareTo(other.GetColorBody().toString());
			
		}
		return 0;
	}
	public boolean equals(Shark other)
	         {
	            if (other == null)
	             {
	                 return false;
	             }
	             if (GetSpeed() != other.GetSpeed())
	             {
	                 return false;
	             }
	             if (GetAge() != other.GetAge())
	             {
	                 return false;
	             }
	             if (GetWeight() != other.GetWeight())
	             {
	                 return false;
	             }
	             if (GetColorBody() != other.GetColorBody())
	             {
	                 return false;
	             }
	             return true;
	         }
	@Override
	         public  boolean equals(Object obj)
	         {
	             if (obj == null)
	             {
	                 return false;
	             }
	             Shark sharkObj = (Shark)obj;
	             if (sharkObj == null)
	             {
	                 return false;
	             }
	             else
	             {
	                 return equals(sharkObj);
	             }
	         }
}
